tachyon/math/base/semigroups.h

```diff
@@ -216,6 +324,107 @@ class AdditiveSemigroup {
+        ret[j] = bases[j].ScalarMul(scalar.ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalars: [s₀, s₁, ..., sₙ₋₁]
+  // base: G
+  // return: [s₀G, s₁G, ..., sₙ₋₁G]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchScalarsWithBase(
+      const std::vector<F>& scalars, const G& base) {
+    size_t size = scalars.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(scalars);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = base.ScalarMul(scalars[j].ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalars: [s₀, s₁, ..., sₙ₋₁]
+  // bases: [G₀, G₁, ..., Gₙ₋₁]
+  // return: [s₀G₀, s₁G₁, ..., sₙ₋₁Gₙ₋₁]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchScalarsToMultiBases(
+      const std::vector<F>& scalars, std::vector<G>& bases) {
+    CHECK_EQ(scalars.size(), bases.size());
+    size_t size = scalars.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(scalars);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = bases[j].ScalarMul(scalars[j].ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalar_powers: [s, s², .., sⁿ⁻¹]
+  // base: G
+  // return: [G, sG, s²G, ..., sⁿ⁻¹G]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchPowers(size_t size, const F& scalar,
+                                           const G& base) {
+// TODO(insun35): Optimize BatchPowers to compute ScalarMul inside of the
+// parallel loop simultaneously without using BatchScalarsWithBase.
+#if defined(TACHYON_HAS_OPENMP)
+    std::vector<F> scalars = F::GetSuccessivePowers(size, scalar);
+    return BatchScalarsWithBase(scalars, base);
+#else
+    return BatchPowersSerial<F>(size, base, scalar);
+#endif
+  }
+
+ private:
+  template <typename ContainerType>
+  static inline size_t GetNumElementsPerThread(const ContainerType& container) {
+#if defined(TACHYON_HAS_OPENMP)
+    size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
+#else
+    size_t thread_nums = 1;
+#endif
+    size_t size = container.size();
+    return std::max(size / thread_nums, size_t{1024});
+  }
+
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchPowersSerial(size_t size, const G& base,
```

Comment:
How about MulBatchPowersSerial? I think "Batch" shouldn't be in the function name in this case.

Reply:
We decided to remove `BatchPowers` at the moment, cuz there are not many use cases using this logic.